codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestEncoder.java

```diff
@@ -37,37 +39,42 @@ public boolean acceptOutboundMessage(Object msg) throws Exception {
     @Override
     protected void encodeInitialLine(ByteBuf buf, HttpRequest request) throws Exception {
         ByteBufUtil.copy(request.method().asciiName(), buf);
-        buf.writeByte(SP);
 
         // Add / as absolute path if no is present.
         // See http://tools.ietf.org/html/rfc2616#section-5.1.2
         String uri = request.uri();
 
         if (uri.isEmpty()) {
-            uri += SLASH;
+            // write " / "
+            ByteBufUtil.writeMediumBE(buf, SPACE_SLASH_AND_SPACE_MEDIUM);
         } else {
+            boolean needSlash = false;
             int start = uri.indexOf("://");
             if (start != -1 && uri.charAt(0) != SLASH) {
-                int startIndex = start + 3;
+                start += 3;
                 // Correctly handle query params.
                 // See https://github.com/netty/netty/issues/2732
-                int index = uri.indexOf(QUESTION_MARK, startIndex);
+                int index = uri.indexOf(QUESTION_MARK, start);
                 if (index == -1) {
-                    if (uri.lastIndexOf(SLASH) <= startIndex) {
-                        uri += SLASH;
+                    if (uri.lastIndexOf(SLASH) <= start) {
+                        needSlash = true;
                     }
                 } else {
-                    if (uri.lastIndexOf(SLASH, index) <= startIndex) {
+                    if (uri.lastIndexOf(SLASH, index) <= start) {
                         uri = new StringBuilder(uri).insert(index, SLASH).toString();
```

Comment:
we could avoid doing the `toString()` conversion here as we will write a `CharSequence` below.